cpp/src/arrow/dataset/file_csv.cc

```diff
@@ -177,20 +177,15 @@ static inline Result<csv::ReadOptions> GetReadOptions(
 }
 
 static inline Future<std::shared_ptr<csv::StreamingReader>> OpenReaderAsync(
-    const FileSource& source, const CsvFileFormat& format,
-    const std::shared_ptr<ScanOptions>& scan_options, Executor* cpu_executor) {
+    const std::shared_ptr<io::BufferedInputStream>& input, const std::string& path,
```

Comment:
Why not take `InputStream` instead?

Reply:
This is private interface, and we know the input is BufferedInputStream, no reason to support polymorphism here I think.